tasks/base_tasks.py

```diff
@@ -1592,3 +1610,133 @@ def collect_meta_wrappers(test_module=None, test_all=True):
             yield t, params
             if not test_all:
                 break
+
+
+class CreateRepoTable(Task):
+    schema = 'repo'
+    table = 'filerepo'
+
+    def run(self):
+        session = current_session()
+
+        create_schema = '''
+                        CREATE SCHEMA IF NOT EXISTS "{schema}"
+                        '''.format(schema=self.output().schema)
+        session.execute(create_schema)
+
+        create_table = '''
+                        CREATE TABLE IF NOT EXISTS "{schema}".{table} (
+                            id TEXT,
+                            version NUMERIC,
+                            checksum TEXT,
+                            url TEXT,
+                            path TEXT,
+                            added TIMESTAMP,
+                            PRIMARY KEY(id, version)
+                        );
+                        '''.format(schema=self.output().schema,
+                                   table=self.output().tablename)
+        session.execute(create_table)
+        session.commit()
+
+    def output(self):
+        return PostgresTarget(schema=self.schema, tablename=self.table, non_empty=False)
+
+
+class RemoteDownloader:
+    def download(self, url, output_path):
+        raise NotImplementedError
+
+
+class BaseDownloader(RemoteDownloader):
+    def download(self, url, output_path):
+        urllib.request.urlretrieve(url, output_path)
+
+
+class RepoFile(Task):
+    resource_id = Parameter()
+    url = Parameter()
+    version = IntParameter()
+    downloader = Parameter(default=BaseDownloader())
+
+    _repo_dir = 'repo'
+    _path = None
+
+    def requires(self):
+        return CreateRepoTable()
+
+    def run(self):
+        self.output().makedirs()
+        digest = self._retrieve_remote_file()
+        self._to_db(self.resource_id, self.version, digest, self.url, self.output().path)
+
+    def _create_filepath(self):
+        return self._build_path(str(uuid.uuid4()))
+
+    def _build_path(self, filename):
+        return os.path.join(self._repo_dir, self.resource_id, str(self.version), filename)
+
+    def _retrieve_remote_file(self):
```

Comment:
Minor: making `_retrieve_remote_file` directly return the digest feels random, but, as it's private, feel free to ignore.

Reply:
You are right. When I created the function I didn't want to introduce another indirection but taking a look after a while it feels a bit weird. I would glad to discuss a change if you think it's needed.

Reply:
Somehow following the command-query separation principle I'd remove `return self._digest(self.output().path)` from `_retrieve_remote_file` and change

```python
digest = self._retrieve_remote_file()
```

into

```python
self._retrieve_remote_file()
digest = self._digest(self.output().path)
```

Although it introduces sequence coupling (you must run `_digest` after `_retrieve_remote_file`), I think that it makes sense from an OO and Python point of view. And, in addition, that coupling is there anyway, as `self.output().path` is passed to `_to_db`.

Reply:
Totally agree. I couldn't see the forest for the trees. Changed it